### ISSUE TITLE
Cosmetic filtering not enforced at uBO launch on already opened web pages

### ISSUE THREAD

**Issue by LittleVulpix** [uBlockOrigin/uBlock-issues#403]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

Blocking early requests no longer works on Waterfox with v1.18.0 . I see there was a recent commit that moved the feature to normal (from experimental) due to new api in Firefox, but since waterfox is a few versions behind, it does not have this API.

EDIT:
I tested it just now with the latest "Legacy" uBO from github and it works properly again. Guess I should be using that one. Not sure if this is therefore even worth the report... perhaps some version detection to not use said api if not available? May be too much hassle, I understand. Even so, here it is. Feel free to close the issue if you think it is not relevant.

### A specific URL where the issue occurs

Everywhere where you have a blocked element.

### Steps to Reproduce

Described in detail here https://github.com/gorhill/uBlock/issues/1327 .

Basically a startup page browser opens with is not filtered.

### Expected behavior:

Blocked objects are properly blocked on browser startup

### Actual behavior:

Blocked objects are not blocked on browser startup

### Your environment

* uBlock Origin version: v1.18.0
* Browser Name and version: Waterfox 56.2.6
* Operating System and version: W10 x86_64, 1703, latest patches.

**Comment by LittleVulpix** [comment]
I had a look through https://github.com/MrAlex94/Waterfox/commits/master and I cannot find the 1447551 or 1503721 referenced, so I'm thinking not, which is why I suspected it broke with latest uBO

**Comment by gorhill** [comment]
Waterfox is not a supported browser.

**Comment by LittleVulpix** [comment]
Ah. That is sad; well, the legacy version works. Thanks for the reply.

**Comment by gorhill** [comment]
Actually, looking at the new code for blocking early request, I am not sure why there would be a change in behavior for FF56. The early blocking is now unconditional, and even without the bug fix in Firefox 61, the early requests should all be held back until uBO is ready.

The difference from before is that the early requests were cancelled (which is what still happen on Chromium), while now they are held back and properly evaluated when uBO is fully loaded. The Firefox fix was to ensure an extension's listener is registered as early as possible, which was an issue for either ways of blocking early requests.

So I wonder if your issue is more about cosmetic filters, since I can't figure how Firefox 56 would have its behavior changed with regard to early network requests.

**Comment by LittleVulpix** [comment]
I posted it previously -> https://github.com/gorhill/uBlock/issues/3451 ; it's the same issue. So I suppose it regards cosmetic filters. Back then, I was pointed to the "suspend tabs until ready" preference, which fixed the problem for me. Until v1.18.0 anyway.

**Comment by gorhill** [comment]
`suspendTabsUntilReady` forces a reload of the page, and this no longer happens now on Firefox with the new code.

If content scripts are not injected (content scripts are required for cosmetic filtering to work), then this means there is a chance stable Firefox would also be affected. However I created a test case and the filters, network and cosmetic, worked as expected:

- Add to "My filters":
    - `||news.ycombinator.com/grayarrow.gif$image`
    - `news.ycombinator.com##.sitestr
- Set `https://news.ycombinator.com/` as Home page
- Quit Firefox
- Launch Firefox

Result: both the arrow gif and the origin strings are removed for each entry on the page, as expected.

**Comment by gorhill** [comment]
If time allow I will try my test case above on FF56.

**Comment by LittleVulpix** [comment]
I don't understand anything anymore. After I downgraded to legacy and now upgraded back to v1.18.0, it no longer occurs. It was happening entire yesterday (had to manually refresh the page on startup to make filtering work). But now... it works. On the test page you mentioned but also on my start page where it was previously broken. Because it did work after page reload then, I ruled out change of the element name, cause a reload would do nothing then.

Again, sorry. No idea what is happening. I don't know if "reinstalling" uBO did anything; I would imagine not. But it works now.

**Comment by gorhill** [comment]
From the look of your screenshot, this is a cosmetic filter issue, hence a content script injection issue. It's possible that the content scripts not being injected also affect Firefox stable. The browser is responsible for injecting the content scripts, and for some reasons as per your screenshot, this did not happen -- though network requests were blocked as per filter.

**Comment by LittleVulpix** [comment]
> 
> 
> `suspendTabsUntilReady` is experimental, so it may not work sometimes.

It worked until v1.18.0, I assume by triggering a page reload as @gorhill said. (which I noticed a few times, but frankly I'm ok with a page being reloaded automatically than seeing ads and things I manually filtered out).

And yeah, this is a widget content which loads into the page as per your preferences. Our amazing primary search provider decided it was a good idea to be unable to remove some of the widgets and also to occassionally add some random ones they thought I'd be interested in. I filtered the whole section out, but with v1.18.0 it doesn't work anymore (not always anyway).

I noticed the page reload never happens either. I'm guessing because the `suspendTabsUntilReady `is gone now as we rely on some new API from the new firefox?

**Comment by gorhill** [comment]
The auto reload was hiding the problem. There is code I added a while ago to manually inject content scripts at launch in case the browser didn't do it: <https://github.com/gorhill/uBlock/blob/c3b0fd31f64bd7ffecdd282fb1208fe07aac3eb0/src/js/start.js#L86>.

So apparently this code is not working on your side, but it seems to be working on my side. So I wonder if this is caused by a fix in Firefox which did not make it into Waterfox.

**Comment by LittleVulpix** [comment]
A thing to note perhaps is that it was not waterfox that was updated; it was uBO. I.e. nothing changed @ Waterfox side, I've been using this version for about a month and it worked fine. It only broke when uBO updated. I have a full backup of my old profile and when I opened it, I found that I was using version

 "version": "1.17.4",

prior to updating to v1.18.0. The release date of 1.17.4 would suggest it already contained the aforementioned change though, so that's odd...

**Comment by gorhill** [comment]
> nothing changed @ Waterfox side

As said, the auto-reload is now gone, so the code above should have picked up the cosmetic filter counterpart of dealing with early filtering.

I am running tests and it seems it's not working on Firefox stable either, but it's working with Chromium.

**Comment by gorhill** [comment]
For reasons unknown, the messaging port in the content script is being disconnected. thus causing the content script to be unable to talk to the main process.

**Comment by gorhill** [comment]
If I delay (using `requestIdleCallback`) the call from the content script to query uBO's main process, the cosmetic filtering is properly applied at launch -- i.e. no spurious port disconnection occur. Of course this is not a solution but this narrows the issue. At this point I am unsure whether the spurious port disconnection is a by-design issue or an actual issue with the webext framework.

**Comment by LittleVulpix** [comment]
Sounds like a pretty fun (and odd) issue to investigate :D Good luck!

**Comment by LittleVulpix** [comment]
@gorhill Hi! The fix works on Waterfox as well. I can see the blocked elements showing for a split second before disappearing - which from I understand is how it should work - the uBO filtering is injected into the page after it loads.

Thanks for the fix!
